clippy_lints/src/empty_docs.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_ast::ast::*;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects documentation that is empty.
+    /// ### Why is this bad?
+    /// It is unlikely that there is any reason to have empty documentation for an item
+    ///
+    /// ### Example
+    /// ```rust
+    /// ///
+    /// fn returns_true() -> bool {
+    ///     true
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn returns_true() -> bool {
+    ///     true
+    /// }
+    /// ```
+    #[clippy::version = "1.74.0"]
+    pub EMPTY_DOCS,
+    suspicious,
+    "docstrings exist but documentation is empty"
+}
+
+declare_lint_pass!(EmptyDocs => [EMPTY_DOCS]);
+
+fn trim_comment(comment: &str) -> String {
+    comment
+        .trim()
+        .split("\n")
+        .map(|comment| comment.trim().trim_matches('*').trim_matches('!'))
+        .filter(|line| !line.is_empty())
+        .collect::<Vec<&str>>()
+        .join("")
+}
+
+impl EarlyLintPass for EmptyDocs {
+    fn check_attribute(&mut self, cx: &EarlyContext<'_>, attribute: &Attribute) {
+        if let AttrKind::DocComment(_line, comment) = attribute.kind {
```

Comment:
It's not the right way to check it. For example in:

```rust
/// a
///
/// b
```

It would emit a warning for the `///`, which is definitely wrong as the doc comment as a whole is not empty. Best way to approach this is instead to check that if an item has doc comments, as a whole, they're not empty.

In case you didn't know, `///` is sugar syntax for `#[doc = ""]`. So your code could be translated as:

```rust
#[doc = "a"]
#[doc = ""]
#[doc = "b"]
```

So to achieve this, you will need to instead make a late lint pass, use [item_attrs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.item_attrs), only keep doc comments and then merge them in one string that you will trim. Then if trimmed string is empty, you can emit the lint.

Reply:
We also already have all the logistics to get the merged docs in `clippy_lints/src/docs/mod.rs`. Making the lint reuse that would be good both from a perf and maintainability perspective.

Reply:
Even better!

Reply:
cool, will look into this.